src/src/com/microsoft/aad/adal/ChallangeResponseBuilder.java

```diff
@@ -85,13 +85,14 @@ public String getAuthorizationHeaderValue() {
      *            client must convey back>
      * @return Return Device challange response
      */
-    public ChallangeResponse getChallangeResponseFromUri(final String redirectUri) {
+    public ChallangeResponse getChallangeResponseFromUri(final String redirectUri)
+            throws UnexpectedServerResponseException {
         ChallangeRequest request = getChallangeRequest(redirectUri);
         return getDeviceCertResponse(request);
     }
 
     public ChallangeResponse getChallangeResponseFromHeader(final String challangeHeaderValue,
-            final String endpoint) throws UnsupportedEncodingException {
+            final String endpoint) throws UnsupportedEncodingException, UnexpectedServerResponseException {
```

Comment:
ResourceAuthenticationChallengeException should be thrown from the AuthenticationParameters class.  That class is about parsing challenges that come from resources.  This class is about parsing device auth challenges that come from the authentication server, not the resource.